docs/spec/lightclient/detection/req-ibc-detection.md

```diff
@@ -0,0 +1,296 @@
+  different branch than the relayer. 
+  And we would like that this is enough to achieve
+  the Handler requirement.
+  
+   - here the correctness argument would be easy if a correct relayer is
+     based on a light client with a *trusted* state, that is, a light
+     client who never changes its opinion about trusted. Then if such a
+     correct relayer checks-in with a handler, it will detect a fork, and
+     act in time.
+
+   - if the light client does not provide this interface, in the case of
+     a fork, we need some assumption about a correct relayer being on a
+     different branch than the handler, and we need such a relayer to 
+	 check-in not too late. Also
+     what happens if the relayer's light client is forced to roll-back
+     its lightstore?
+     Does it have to re-check all handlers?
+
+
+
+
+## On the interconnectedness of things
+
+In the broader discussion of so-called "fork accountability" there are
+several subproblems
+
+- Fork detection 
+
+- Evidence creation and submission
+
+- Isolating misbehaving nodes (and report them for punishment over abci)
+
+
+### Fork detection
+
+The preliminary specification ./detection.md formalizes the notion of
+a fork. Roughly, a fork exists if there are two conflicting headers
+for the same height, where both are supported by bonded full nodes
+(that have been validators in the near past, that is, within the
+trusting period). We distinguish between *fork on the chain* where two
+conflicting blocks are signed by +2/3 of the validators of that
+height, and a *light client fork* where one of the conflicting headers
+is not signed by  +2/3 of the current height, but by +1/3 of the
+validators of some smaller height.
+
+In principle everyone can detect a fork
+
+- ./detection talks about the Tendermint light client with a focus on 
+  light nodes.
+  
+- a relayer runs such light clients and may detect forks in this way
+
+- in IBC, a relayer can see that a handler is on a conflicting branch
+    - the relayer should feed the handler the necessary information so
+      that it can halt
+	- the relayer should report the fork to a full node
+	
+### Evidence creation and submission
+
+- the information sent from the relayer to the handler could be called
+  evidence, but this is perhaps a bad idea because the information sent to a
+  full node can also be called evidence. But this evidence might still
+  not be enough as the full node might need to run the "fork
+  accountability" protocol to generate evidence in the form of
+  consensus messages. So perhaps we should
+  introduce different terms for:
+  
+  - proof of fork for the handler (basically consisting of lightblocks)
+  - proof of fork for a full node (basically consisting of (fewer) lightblocks)
+  - proof of misbehavior (consensus messages)
+  
+
+
+### Isolating misbehaving nodes
+
+- this is the job of a full node. 
+
+- might be subjective in the future: the protocol depends on what the
```

Comment:
In Tendermint we assume that we have correct gossip paths between all correct validators. This allows information about fork observed by one correct validator to be eventually received by all correct validators. Therefore eventually all correct validators panic and we halt chain. On the other side, we don't make explicit assumptions regarding full nodes and connectivity between correct full nodes. So it is possible having a full node that is being eclipsed by adversary on the forked chain. The question is if this change something in protocol design when we talk about evidence submission?Light client will anyway need to communicate proof of fork to primary and all witnesses so we ensure that this information reach out correct full nodes. Note that our assumption of light client being connected to a correct peer is actually not precise enough. We probably need to say that it is connected to a correct full node that is on the main chain.

Reply:
I am also not clear about the right way to think about that. I just wanted to highlight that currently all our discussion implicitly assume that a full node that tries to collect evidence knows the "true" chain, and that I am not sure we want to keep it that way. It might also be that anyway we have to halt the system in such situations...